frameworks/Rust/actix/src/main.rs

```diff
@@ -2,161 +2,53 @@ extern crate actix;
 extern crate bytes;
 extern crate futures;
-extern crate num_cpus;
-extern crate rand;
 extern crate serde;
 extern crate serde_json;
 #[macro_use]
 extern crate serde_derive;
-#[macro_use]
-extern crate diesel;
-#[macro_use]
-extern crate horrorshow;
 
 use actix::prelude::*;
-use actix_web::{http, server, App, AsyncResponder, FutureResponse, HttpRequest,
-                HttpResponse};
+use actix_web::{http, server, App, HttpRequest, HttpResponse};
 use bytes::BytesMut;
-use diesel::prelude::{Connection, PgConnection};
-use futures::Future;
-use std::cmp;
 
-mod db;
-mod models;
-mod schema;
-mod tmpl;
 mod utils;
 use utils::Writer;
 
-struct State {
-    db: Addr<Syn, db::DbExecutor>,
-}
-
-fn world_row(req: HttpRequest<State>) -> FutureResponse<HttpResponse> {
-    req.clone()
-        .state()
-        .db
-        .send(db::RandomWorld)
-        .from_err()
-        .and_then(move |res| match res {
-            Ok(row) => {
-                let mut body = BytesMut::with_capacity(31);
-                serde_json::to_writer(Writer(&mut body), &row).unwrap();
-                Ok(HttpResponse::build_from(&req)
-                    .header(http::header::SERVER, "Actix")
-                    .content_type("application/json")
-                    .body(body))
-            }
-            Err(_) => Ok(HttpResponse::InternalServerError().into()),
-        })
-        .responder()
-}
-
-fn queries(req: HttpRequest<State>) -> FutureResponse<HttpResponse> {
-    // get queries parameter
-    let q = if let Some(q) = req.query().get("q") {
-        q.parse::<u16>().ok().unwrap_or(1)
-    } else {
-        1
-    };
-    let q = cmp::min(500, cmp::max(1, q));
+const SIZE: usize = 29;
 
-    // run sql queries
-    req.clone()
-        .state()
-        .db
-        .send(db::RandomWorlds(q))
-        .from_err()
-        .and_then(move |res| {
-            if let Ok(worlds) = res {
-                let mut body = BytesMut::with_capacity(35 * worlds.len());
-                serde_json::to_writer(Writer(&mut body), &worlds).unwrap();
-                Ok(HttpResponse::build_from(&req)
-                    .header(http::header::SERVER, "Actix")
-                    .content_type("application/json")
-                    .body(body))
-            } else {
-                Ok(HttpResponse::InternalServerError().into())
-            }
-        })
-        .responder()
+#[derive(Serialize, Deserialize)]
+pub struct Message {
+    pub message: &'static str,
 }
 
-fn updates(req: HttpRequest<State>) -> FutureResponse<HttpResponse> {
-    // get queries parameter
-    let q = if let Some(q) = req.query().get("q") {
-        q.parse::<usize>().ok().unwrap_or(1)
-    } else {
-        1
+fn json(req: HttpRequest) -> HttpResponse {
+    let message = Message {
+        message: "Hello, World!",
     };
-    let q = cmp::min(500, cmp::max(1, q));
+    let mut body = BytesMut::with_capacity(SIZE);
+    serde_json::to_writer(Writer(&mut body), &message).unwrap();
```

Comment:
Is there a reason to not use `serde_json::to_string(&message).unwrap()` here?

Reply:
It will allocate String, then you’ll need to convert it to Bytes. In this code serde writes to Bytes directly

Reply:
I wonder if zero-cost abstractions of rust helped in this such that they know we want the bytes after conversion to string and if it could skip the process directly. Nevertheless, it seems weird that serde have `to_string` but not `to_bytes`. Let me create an issue there.

Reply:
You can convert String to Vec, which is zero cost. Serde can use io::Write and this code uses it

Reply:
Yeah, but that works but not ergonomic. I wonder if `serde_json::to_string('string').as_bytes()` zero cost?

Reply:
it is not in sense it is allocates.